### PR TITLE
Add support for inline JSDoc links

### PR DESCRIPTION
**Pull request by Alhadis**
This PR adds pattern-matching for two JSDoc tags, both of which share the same syntax:

* [`@link`](http://usejsdoc.org/tags-inline-link.html)
* [`@tutorial`](http://usejsdoc.org/tags-inline-tutorial.html)

**Preview:**  
Figure 1 Figure 2

### CODE REVIEW

File: grammars/javascript.cson [hunk 1]
```diff
@@ -1796,13 +1796,62 @@
...
+                  | tutorial
+                )
+              )
+              
+              \\s+
+
+              ([^\\s|}]+)        # Namepath or URL
+
+              (?:                # Optional link text following link target
+                [\\s|]           # Bar or space separating target and text
```
Comment (50Wliu) [inline grammars/javascript.cson]: Can there be multiple spaces?  Or is any space after the first considered part of the text?

File: grammars/javascript.cson [hunk 2]
```diff
@@ -1796,13 +1796,62 @@
...
+                [^\\}]*          # Actual text
```
Comment (50Wliu) [inline grammars/javascript.cson]: `[^}]*`

File: spec/javascript-spec.coffee [hunk 3]
```diff
@@ -1563,6 +1563,52 @@ describe "Javascript grammar", ->
...
+      expect(tokens[5]).toEqual value: 'target', scopes: ['source.js', 'comment.block.documentation.js', 'other.meta.jsdoc', 'entity.name.type.instance.jsdoc', 'variable.other.description.jsdoc']
+      expect(tokens[6]).toEqual value: ' Description text}', scopes: ['source.js', 'comment.block.documentation.js', 'other.meta.jsdoc', 'entity.name.type.instance.jsdoc']
+      expect(tokens[8]).toEqual value: '*/', scopes: ['source.js', 'comment.block.documentation.js', 'punctuation.definition.comment.js']
+
+      {tokens} = grammar.tokenizeLine('/** Text [Description text]{@link target} */')
+      expect(tokens[0]).toEqual value: '/**', scopes: ['source.js', 'comment.block.documentation.js', 'punctuation.definition.comment.js']
+      expect(tokens[1]).toEqual value: ' Text ', scopes: ['source.js', 'comment.block.documentation.js']
+      expect(tokens[2]).toEqual value: '[Description text]{', scopes: ['source.js', 'comment.block.documentation.js', 'other.meta.jsdoc', 'entity.name.type.instance.jsdoc']
```
Comment (50Wliu) [inline spec/javascript-spec.coffee]: I don't think the `{` should have the same scopes as `[Description text]`.

File: spec/javascript-spec.coffee [hunk 4]
```diff
@@ -1563,6 +1563,52 @@ describe "Javascript grammar", ->
...
+
+      {tokens} = grammar.tokenizeLine('/** Text [Description text]{@tutorial target|Description} */')
+      expect(tokens[0]).toEqual value: '/**', scopes: ['source.js', 'comment.block.documentation.js', 'punctuation.definition.comment.js']
+      expect(tokens[1]).toEqual value: ' Text [Description text]', scopes: ['source.js', 'comment.block.documentation.js']
+      expect(tokens[2]).toEqual value: '{', scopes: ['source.js', 'comment.block.documentation.js', 'other.meta.jsdoc', 'entity.name.type.instance.jsdoc']
+      expect(tokens[3]).toEqual value: '@tutorial', scopes: ['source.js', 'comment.block.documentation.js', 'other.meta.jsdoc', 'entity.name.type.instance.jsdoc', 'storage.type.class.jsdoc']
+      expect(tokens[5]).toEqual value: 'target', scopes: ['source.js', 'comment.block.documentation.js', 'other.meta.jsdoc', 'entity.name.type.instance.jsdoc', 'variable.other.description.jsdoc']
+      expect(tokens[6]).toEqual value: '|Description}', scopes: ['source.js', 'comment.block.documentation.js', 'other.meta.jsdoc', 'entity.name.type.instance.jsdoc']
```
Comment (50Wliu) [inline spec/javascript-spec.coffee]: Ditto

File: grammars/javascript.cson (hunk 1 above)
Comment (Alhadis) [inline grammars/javascript.cson]: The former. The latter requires a pipe: `{@link http://www.google.com|  My link}`.

File: spec/javascript-spec.coffee (hunk 4 above)
Comment (Alhadis) [inline spec/javascript-spec.coffee]: BTW, you'll be pleased to know I won't be including any helper functions in specs from now on. =) I just [generate the CoffeeScript lines](https://github.com/Alhadis/Atom-PhoenixTheme/commit/37d7e6d9be8b94ce95986fad403976a064d43d3b) instead.

File: spec/javascript-spec.coffee (hunk 3 above)
Comment (50Wliu) [inline spec/javascript-spec.coffee]: `Description text` probably deserves its own scopes.  Maybe not give dedicated punctuation scopes yet, but that can definitely be investigated in a different PR.

Comment (Alhadis) [inline spec/javascript-spec.coffee]: I've added `constant.other.description.jsdoc`. It isn't a constant, but there's really nothing semantically-correct I could add that doesn't use the `entity` scope.

**Comment by 50Wliu** [comment]
Rebase required.

**Comment by Alhadis** [comment]
Sorted.
